toolchain/common/index_base.h

```diff
@@ -0,0 +1,48 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_COMMON_INDEX_BASE_H_
+#define CARBON_TOOLCHAIN_COMMON_INDEX_BASE_H_
+
+#include <type_traits>
+
+#include "common/ostream.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/iterator.h"
+#include "llvm/Support/Format.h"
+
+namespace Carbon {
+
+template <typename DataType>
+class DataIterator;
+
+// A lightweight handle to an item in a vector.
+//
+// DataIndex is designed to be passed by value, not reference or pointer. They
+// are also designed to be small and efficient to store in data structures.
+struct IndexBase {
+  IndexBase() : index(-1) {}
+  explicit IndexBase(int index) : index(index) {}
+
+  auto Print(llvm::raw_ostream& output) const -> void { output << index; }
+
+  int32_t index;
+};
```

Comment:
Any particular reason to not make this convertible, that would seem more convenient to me?

I also personally would find it a bit less surprising when reading users of this to have this be private and use an accessor method (maybe in addition to an implicit conversion for use with `[]`s). Any reason you prefer this?

Reply:
Style seems to make a strong statement here, "Do not define implicit conversions." https://google.github.io/styleguide/cppguide.html#Implicit_Conversions

"Implicit conversions can sometimes be necessary and appropriate for types that are designed to be interchangeable, for example when objects of two types are just different representations of the same underlying value." -- but I'm not convinced that these should be treated as interchangeable. The cost of `.index` feels low to me.

e.g., I think the line math is awkward and shouldn't magically/transparently work. Although really, maybe the line value should be a number unlike everything else, representing how it's really the line number, not some abstract index.

Reply:
Probably notable too that I would still lean towards _more_ type safety, not less.